### PR TITLE
SMP Support RISC-V PolarFire SoC Icicle Kit

### PR DESCRIPTION
**Pull request by con-pax**
The first cut of the Risc-V SMP support provides basic support for machine mode SMP operation with integer and optional floating point context saving via the arch_switch mechanism. The configuration options allow selection of a contiguous set of CPUs between 1 and 4 to allow fine grained selection of a subset of the available CPUs. Non-homogeneous CPU setups such as those found in the Microchip PolarFire SOC and Sifive FU540 with a mix of monitor and application CPUs, are supported by allowing the monitor CPU be excluded from the operating set.

### CODE REVIEW

**Comment by edersondisouza** [comment]
Hi @con-pax,

It's great to see support being added for the PolarFire SoC Icicle Kit! As you know, I've sent #41939 that also touches SMP for RISC-V, but it's qemu only. I hope you've had a chance to take a look at it =D

Let's talk on Zephyr's Discord about ways to converge both efforts? My nick there is also "edersondisouza" - not sure about your timezone, but I'm there usually from 17:00 UTC to 2:00 UTC - but we can certainly find the best time to chat =D

Cheers!

**Comment by con-pax** [comment]
Hi @nashif,
Thanks for you feedback, I will reformat the commits, and ensure that the commit messages meet the requirement, then try again.

Hi @edersondisouza, Thanks. that sounds like a great idea! I will come find you on the Discord and we can go from there.

File: soc/riscv/riscv-privilege/mpfs/mpfs_sys.c [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		    0,
+		    icicleIRQHandler,
+		    NULL,
+		    0);
+		    
+       MSS_GPIO_enable_irq(GPIO2_LO, MSS_GPIO_30);
+       MSS_GPIO_enable_irq(GPIO2_LO, MSS_GPIO_31);
+    
+	/* Enable IRQ for PLIC driver */
+	//irq_enable(RISCV_MACHINE_EXT_IRQ);
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/mpfs_sys.c:132]: Unused commented-out code?

File: soc/riscv/riscv-privilege/mpfs/mpfs_sys.h [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+/* */
+#define SOFT_RESET_CR_MMUART4_OFFSET                         0x9
+#define SOFT_RESET_CR_MMUART4_MASK                           (0x01 << 0x9)
+
+    #define SUBBLK_CLOCK_CR_MMUART0_OFFSET                       0x5
+    #define SUBBLK_CLOCK_CR_MMUART0_MASK                         (0x01 << 0x5)
+    #define SUBBLK_CLOCK_CR_MMUART1_MASK                         (0x01 << 0x6)
+#define SUBBLK_CLOCK_CR_GPIO0_MASK                           (0x01 << 0x14)
+    #define SUBBLK_CLOCK_CR_GPIO2_MASK                           (0x01 << 0x16)
+        #define SUBBLK_CLOCK_CR_CFM_MASK                             (0x01 << 0x1D)
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/mpfs_sys.h:38]: Indentation

File: soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    return (0U);
+}
+
+static int icicle_before_init(const struct device *dev)
+{
+
+#ifdef  MPFS_HAL_HW_CONFIG
+        load_virtual_rom();
+//        config_l2_cache();
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:84]: Unused code

File: soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+#endif  /* MPFS_HAL_HW_CONFIG */
+
+        //init_memory
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:87]: Ditto

File: soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        __disable_all_irqs();      /* disables local and global interrupt enable */
+
+#ifdef  MPFS_HAL_HW_CONFIG
+       (void)init_bus_error_unit(); //not addded
+//       (void)init_mem_protection_unit();
+//       (void)init_pmp((uint8_t)MPFS_HAl_FIRST_HART);
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:93]: Ditto

File: soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+#endif
+ 
+       PLIC_ClearPendingIRQ();
+//        mss_nwc_init();
+//        mss_pll_config();
+        
+
+        //PLIC_init_on_reset();
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:101]: Ditto

**Comment by con-pax** [comment]
> You need to work on many of those commits and commit messages, squash many of them and reduce this to a small set of commits for each area...

Hi @nashif,

Wondering if I should close this PR, and create a new one from a branch with the rebased/squashed commits?
Kind Regards,
Conor

**Comment by nashif** [comment]
> Wondering if I should close this PR, and create a new one from a branch with the rebased/squashed commits?
> Kind Regards,

you do not need to close, you can squash and cleanup commits and force push to the same PR.

File: arch/riscv/core/reset.S [hunk 7]
```diff
@@ -58,36 +135,100 @@ boot_master_core:
...
+	sd a1, 0(a2)
+	li a0, RV_WAKE_GO
+#endif	/* CONFIG_SMP */
+	/*
+	 * Enable software interrupt
+	 */
+	li  t0, MIP_MSIP
+	csrrs x0, mie, t0
 
+loop_slave_core:
```
Comment (kgugala) [inline arch/riscv/core/reset.S:203]: i don't think we should use `master`/`slave` nomenclature in the new code. Maybe `main`/`secondary` core?

File: arch/riscv/core/riscv_smp.c [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+void z_riscv_secondary_start(int cpu_num)
+{
+	arch_cpustart_t fn;
+#if 0
+#ifdef CONFIG_SMP
+	z_icache_setup();
+	z_irq_setup();
+
+#endif
+#endif
```
Comment (kgugala) [inline arch/riscv/core/riscv_smp.c:98]: commented out code

File: arch/riscv/core/riscv_smp.c [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+	riscv_cpu_init[cpu_num].arg = arg;
+
+	/* set the initial sp of target sp through riscv_cpu_sp
+	 * Controlled sequencing of start up will ensure
+	 * only one secondary cpu can read it per time 
+	 */
+	riscv_cpu_sp = Z_THREAD_STACK_BUFFER(stack) + sz;
+
+	/* wait slave cpu to start */
+	while (hart_wake_flags[hart_num] != RV_WAKE_WAIT) {
```
Comment (kgugala) [inline arch/riscv/core/riscv_smp.c:71]: what if the secondary core fails to start (for any reason)? Shall we handle such case?

File: arch/riscv/core/riscv_smp.c [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+    return 0;
+}
+
+SYS_INIT(riscv_smp_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+#endif
+
+
+#if 0
```
Comment (kgugala) [inline arch/riscv/core/riscv_smp.c:152]: a lot of commented out code, please remove it

**Review by kgugala — CHANGES_REQUESTED** [review]
looks like this PR is introducing a few changes at once (SMP and new board) making it quite big piece of code to review. Can you split this into smaller PRs - one touching generic RV code and a followup changing the board code?

File: arch/riscv/core/reset.S [hunk 11]
```diff
@@ -33,18 +39,89 @@ SECTION_FUNC(reset, __reset)
  * the C domain
  */
 SECTION_FUNC(TEXT, __initialize)
-	/*
-	 * This will boot master core, just halt other cores.
-	 * Note: need to be updated for complete SMP support
+    la ra, __initialize /* Keeps debugger happy so it does not follow false trails... */
```
[the remark this answers is not in the document]
Comment (con-pax) [reply arch/riscv/core/reset.S:42]: > looks like this PR is introducing a few changes at once (SMP and new board) making it quite big piece of code to review. Can you split this into smaller PRs - one touching generic RV code and a followup changing the board code?

Hi @kgugala, thanks for taking the time to look at the pr. Yes, the plan is rebase against latest version and split into smaller pr's.

**Review by PiotrZierhoffer — CHANGES_REQUESTED** [review]
Some minor remarks - I tried to compile things and failed, and these are the places I had to address locally.

File: arch/riscv/core/riscv_smp.c [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+		__ASSERT(0,
+			"ARC connect has no global free running counter\n");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+SYS_INIT(arc_smp_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+#endif
```
Comment (evgeniy-paltsev) [inline arch/riscv/core/riscv_smp.c:217]: This looks like copy-paste of ARC SMP code. Is it really used for risc-V?  :)

File: subsys/shell/modules/kernel_service.c [hunk 13]
```diff
@@ -174,9 +174,13 @@ static void shell_stack_dump(const struct k_thread *thread, void *user_data)
 		      size, unused, size - unused, size, pcnt);
 }
 
+if (CONFIG_MP_TOTAL_NUM_CPUS > CONFIG_MP_NUM_CPUS)
```
[the remark this answers is not in the document]
Comment (con-pax) [reply subsys/shell/modules/kernel_service.c:177]: > 

Hi Piotr, thanks for taking the time to look at the PR. There has been a substantial change to the Zephyr code base, in particular the isr.S over the last couple of days. This PR is becoming quite stale. We have started to break up the work into smaller PR's, starting with board support.

[the remark this answers is not in the document]
Comment (PiotrZierhoffer) [reply subsys/shell/modules/kernel_service.c:177]: Great to hear that! I'd be happy to run it through Renode, so please @ me and @kgugala when you file it!

**Comment by con-pax** [comment]
Closing this PR today, as it has gone stale.
PR #44208 has board support for PolarFire SoC Icicle kit and has passed all CI testing.
@kgugala, @PiotrZierhoffer  could I please request you to review it?
Thanks a million for your help so far

**Pull request closed without merging**